qiskit/circuit/classical/expr/constructors.py

```diff
@@ -561,3 +568,210 @@ def index(target: typing.Any, index: typing.Any, /) -> Expr:
     if target.type.kind is not types.Uint or index.type.kind is not types.Uint:
         raise TypeError(f"invalid types for indexing: '{target.type}' and '{index.type}'")
     return Index(target, index, types.Bool())
+
+
+def _binary_sum(op: Binary.Op, left: typing.Any, right: typing.Any) -> Expr:
+    left, right = _lift_binary_operands(left, right)
+    if (
+        left.type.kind is types.Bool
+        or right.type.kind is types.Bool
+        or left.type.kind is not right.type.kind
+        and types.order(left.type, right.type) is types.Ordering.NONE
+    ):
+        raise TypeError(f"invalid types for '{op}': '{left.type}' and '{right.type}'")
```

Comment:
I might be overthinking this, but the "negative" type assertion makes me slightly nervous - it'd be easy to forget to update this.  What I mean is: we assert that the type _isn't_ `Bool`, as opposed to asserting that the kind _is_ in `(Uint, Float, Duration)` (well, the assertion is the negation of that, but my point is "allowlist" vs "blocklist").

I'm just thinking that the allowlist form fails safer under expansion of the type system (we might loudly [falsely] error, rather than silently [falsely] produce nonsense) - maybe worth doing it that way round?

Reply:
That's entirely fair. I was actually just copying what you'd done previously with `Bool` :3 

But I can change this to an allow list!

Reply:
Done in f780d9a. I was also able to remove the order stuff, since that was a holdover from when we had `Stretch` and `Duration` types.